opnode/node/node.go

```diff
@@ -1,18 +1,236 @@
+	// TODO: bootnodes option (bootstrap discovery of more peers)
+
+	log log.Logger
+
+	// (combined) source to fetch data from
+	l1Source eth.L1Source
+
+	// engines to keep synced
+	l2Engines []*l2.EngineDriver
+
+	l1Downloader l1.Downloader
+
+	ctx   context.Context
+	close chan chan error
+}
+
+func (c *OpNodeCmd) Default() {
+	c.L1NodeAddrs = []string{"http://127.0.0.1:8545"}
+	c.L2EngineAddrs = []string{"http://127.0.0.1:8551"}
 }
 
 func (c *OpNodeCmd) Help() string {
 	return "Run optimism node"
 }
 
 func (c *OpNodeCmd) Run(ctx context.Context, args ...string) error {
+	logger := c.LogCmd.Create()
+	c.log = logger
+	c.ctx = ctx
+
+	if c.Genesis == (GenesisConf{}) {
+		return errors.New("genesis configuration required")
+	}
+
+	l1Sources := make([]eth.L1Source, 0, len(c.L1NodeAddrs))
+	for i, addr := range c.L1NodeAddrs {
+		// L1 exec engine: read-only, to update L2 consensus with
+		l1Node, err := rpc.DialContext(ctx, addr)
+		if err != nil {
+			// HTTP or WS RPC may create a disconnected client, RPC over IPC may fail directly
+			if l1Node == nil {
+				return fmt.Errorf("failed to dial L1 address %d (%s): %v", i, addr, err)
+			}
+			c.log.Warn("failed to dial L1 address, but may connect later", "i", i, "addr", addr, "err", err)
+		}
+		// TODO: we may need to authenticate the connection with L1
+		// l1Node.SetHeader()
+		cl := ethclient.NewClient(l1Node)
+		l1Sources = append(l1Sources, cl)
+	}
+	if len(l1Sources) == 0 {
+		return fmt.Errorf("need at least one L1 source endpoint, see --l1")
+	}
+
+	// Combine L1 sources, so work can be balanced between them
+	c.l1Source = eth.NewCombinedL1Source(l1Sources)
+	l1CanonicalChain := eth.CanonicalChain(c.l1Source)
+
+	c.l1Downloader = l1.NewDownloader(c.l1Source)
+
+	for i, addr := range c.L2EngineAddrs {
+		// L2 exec engine: updated by this OpNode (L2 consensus layer node)
+		backend, err := rpc.DialContext(ctx, addr)
+		if err != nil {
+			if backend == nil {
+				return fmt.Errorf("failed to dial L2 address %d (%s): %v", i, addr, err)
+			}
+			c.log.Warn("failed to dial L2 address, but may connect later", "i", i, "addr", addr, "err", err)
+		}
+		// TODO: we may need to authenticate the connection with L2
+		// backend.SetHeader()
+		client := &l2.EngineClient{
+			RPCBackend: backend,
+			EthBackend: ethclient.NewClient(backend),
+			Log:        c.log.New("engine_client", i),
+		}
+		engine := &l2.EngineDriver{
+			Ctx: c.ctx,
+			Log: c.log.New("engine", i),
+			RPC: client,
+			SyncRef: l2.SyncSource{
+				L1: l1CanonicalChain,
+				L2: client,
+			},
+		}
+		c.l2Engines = append(c.l2Engines, engine)
+	}
+
+	// TODO: maybe spin up an API server
+	//  (to get debug data, change runtime settings like logging, serve pprof, get peering info, node health, etc.)
+
+	c.close = make(chan chan error)
+
+	go c.RunNode()
+
 	return nil
 }
 
+func (c *OpNodeCmd) RunNode() {
+	c.log.Info("Starting OpNode")
+
+	var unsub []func()
+	mergeSub := func(sub ethereum.Subscription, errMsg string) {
```

Comment:
rename to handle unsubscribe? My first impression was that this was supposed to merge subscriptions into a single stream